### ISSUE TITLE
Create a Tool to Automatically Insert Semantic Breaks for Overflowing Lines

### ISSUE THREAD

**Issue by mattt** [sembr/specification#1]
Without tooling, a specification like this one can only be descriptive, like a style guide. As a future enhancement, it would be nice to build a tool that automatically inserts line breaks at semantic boundaries for lines that extend beyond a prescribed width (e.g. 80 columns). Like a cross between `prettier` and `fold`.

**Comment by waldyrious** [comment]
Some years ago I created a very crude proof-of-concept to do precisely this. It can be tested here: https://waldyrious.github.io/semantic-linebreaker/, hosted directly from the code in https://github.com/waldyrious/semantic-linebreaker/.

It is very primitive as the README (and the code) attests to, and there are [issues](https://github.com/waldyrious/semantic-linebreaker/issues) to be handled, but any help would be greatly appreciated and quickly merged.

**Comment by waldyrious** [comment]
(Btw, I'd be happy to move the repo to this organization, if that's desired.)

**Comment by mattt** [comment]
Thanks so much for sharing that, @waldyrious!

I think regular expressions offer a quick and clever approximation of the kind of line-breaking behavior we're looking for. However, I don't think it's feasible to consistently express semantic boundaries with them.

My current thinking is that a complete solution would probably have to apply an algorithm like [Knuth-Plass](https://dl.acm.org/citation.cfm?id=2797091) or [Wadler ("prettier printer")](https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf), feeding in tokens from a [linguistic syntax tree](https://en.wikipedia.org/wiki/Parse_tree).

**Comment by SilasK** [comment]
Are there any updates on this?

**Comment by silopolis** [comment]
That readable project indeed looks nice and promising!
Thanks for sharing 🙏

**Comment by waldyrious** [comment]
> That readable project indeed looks nice and promising!

Agreed! For reference, the actual implementation is [here](https://github.com/bobheadxi/readable/blob/main/plugins/readable/reflow.ts).

**Comment by SilasK** [comment]
Does someone have enough Typescript to deactivate the rest of the formater. https://github.com/bobheadxi/readable/issues/30

**Comment by admk** [comment]
This is an itch that haven't been scratched for a long time!

Semantic line breaking is by nature an NLP problem, so I fine-tuned Bert models as token classifiers to predict line breaks on my text and it works reasonably well. I have thus created a tool that uses these models to insert breaks automatically in your text. CUDA (Linux / Windows) or MPS (Mac) acceleration are supported. Currently it works well for LaTeX and plain text, other markup languages are not tested.

The fine-tuned models can be found here on [Hugging Face](https://huggingface.co/collections/admko/sembr-656d5c3f6f39f15658533a64).

Suggestions for improvements and contributions to features, models, or datasets are all welcomed! Feel free to explore and contribute to the project: https://github.com/admk/sembr.

**Comment by silopolis** [comment]
That's playing in another league!
Very very interesting project indeed... With the amount of lightweight markup content produced these days, support for markdown, asciidoc and restructuredtext would surely be fantastic!

Awesome work 🤩

**Comment by firubat** [comment]
Has there been any advances in the last year? This could be a great tool!

**Comment by albertodiazdorado** [comment]
@admk A formatter does not need to solve an NLP problem, right? Wouldn't a tool following heuristics be capable of enforcing 9x% of the standard? That would be good enough for most of us

(Clarifying that "solving an NLP problem" is not a requirement for a successful tool)

**Comment by alerque** [comment]
For those looking for a Pandoc filter for this, a very high percentage of non-technical prose cases are covered for English and Turkish in [this Lua filter](https://github.com/sile-typesetter/casile/blob/master/pandoc-filters/sentence_lines.lua).

**Comment by mattt** [comment]
Closing this one out. Seven years on, there are several tools that do this well enough in practice, and #22 lists them in the FAQ. Thanks to everyone who shared their work here: @waldyrious, @SilasK, @admk, @alerque, and @arloliu.
